Granular validation

alcuadrado opened the pull request:
This PR closes #120 

I tried adding a test for this at the end of the huge `async.series` call, but I can't get it run. It's ignored by tape. Any idea?

holgerd77 commented on src/index.ts:
```diff
@@ -47,10 +47,21 @@ export interface BlockchainOptions {
+   * This flags indicates if blocks should be validated. See Block#validate for details. Defaults to
+   * the value of `validate`.
+   */
+  validateBlocks?: boolean
```
I am not happy with the redundancy introduced by this sublevel of validation along with the general `validation` flag. On the edge cases (1. `validation` to `true`, both other to `false`, 2. `validation` to `false`, both other to `true`) it's just not really decidable or logically derivable which setting super-sedes and will likely lead to erroneous usage.

My first thinking was... 

1. ...that this would be just worth a breaking change and we should drop the `validate` flag altogether, all cases can be disambibuously covered by this.

Then I thought...

2. on the lower end this could be covered by clearly document by saying which parameter hierarchy takes precedence.

A combination of this would be...

3. To do 2. and along mark the `validate` flag as `DEPRECATED`, to get to some less error prone implementation again on the next breaking release.

What do you think?

alcuadrado replied:
I like your proposal. I didn't want to remove `validate` just yet, as it's a breaking change, as it's a breaking change. But flagging it as deprecated is a step forward. I'll also improve the documentation.

holgerd77 replied:
Could you also add the false value defaults as a remark to the docs? (Did you actually remove this along the way, thought this was already in?)

alcuadrado commented:
@holgerd77 I updated this PR with the following changes:

1. Add a `@deprecated` tag in Blockchain's and BlockchainOptions' `validate` field.
2. Throw in the `Blockchain` constructor if both `validate` and `validatePow`/`validateBlocks` are used.
3. Always set `Blockchain#validate` as `true` and ignore its value.
4. Update the docs.

holgerd77 commented:
@alcuadrado Ahm, you still have to set the `validate` option if it is used, otherwise it is not a deprecation but a removal, or am I getting something wrong? 😛

holgerd77 commented:
Update: ah, I am getting something wrong, just seeing that you are directly assigning this to the specific valdation parameters. 😄

holgerd77 commented on src/index.ts:
```diff
@@ -133,11 +152,20 @@ export default class Blockchain {
-    this.validate = opts.validate === undefined ? true : opts.validate
-    this.ethash = this.validate ? new Ethash(this.db) : null
+    this._validatePow = opts.validatePow !== undefined ? opts.validatePow : !!opts.validate
+    this._validateBlocks = opts.validateBlocks !== undefined ? opts.validateBlocks : !!opts.validate
```
Sorry, thought a about this a bit longer and maybe it's exaggerated: but can we make this with an explicit `if (opts.validate) { ... } else { ... }` clause? I am especially afraid that this `true` default value for the two specific parameters gets too easily lost (if e.g. done by someone else) on the transition, since this is too implicitly side-integrated through `opts.validate`.

alcuadrado replied:
Good idea, looks much better that way

alcuadrado commented:
There was a bug on how `validate` and the default values were handled. I fixed it and updated a test to be compatible with the changes introduced by this PR.
